docs/design-documents/platform/crash-reporting/crash_reporting.md

```diff
@@ -0,0 +1,286 @@
+
+```C
+mbed_error_ctx my_error_ctx;
+//Callback during reboot
+void mbed_error_reboot_callback(mbed_error_ctx *error_context) 
+{
+    printf("Error callback received");
+    //Copy the error context in a local struct for processing it later
+    memcpy(&my_error_ctx, error_context, sizeof(mbed_error_ctx));
+}
+```
+The above function will be called during boot with a pointer to *error_context* structure.
+
+### Retrieving error info after reboot
+The error context captured can be retrieved using mbed_get_reboot_error_info() API. See the below code
+for example usage of that API. In the example below, a status variable reboot_error_detected has been used to track the presence of error context capture.
+
+```C
+mbed_error_ctx error_ctx;
+int reboot_error_detected = 0;
+
+//Callback during reboot
+void mbed_error_reboot_callback(mbed_error_ctx *error_context) 
+{
+    printf("error callback received");
+    reboot_error_detected = 1;
+}
+
+// main() runs in its own thread in the OS
+int main() 
+{
+    if (reboot_error_detected == 1) {
+        if (MBED_SUCCESS == mbed_get_reboot_error_info(&error_ctx)) {
+            printf("\nSuccessfully read error context\n");
+    }
+    //main continues...
+}
+```
+
+### Retrieving fault context after reboot
+The fault context captured can be retrieved using mbed_get_reboot_fault_context() API. See the below code
+for example usage of that API. The example code below checks for error_status using the error context and then
+retrieves the fault context using mbed_get_reboot_fault_context() API.
+
+```C
+mbed_error_ctx error_ctx;
+mbed_fault_context_t fault_ctx;
+int reboot_error_detected = 0;
+
+//Callback during reboot
+void mbed_error_reboot_callback(mbed_error_ctx * error_context) 
+{
+    printf("error callback received");
+    reboot_error_detected = 1;
+}
+
+// main() runs in its own thread in the OS
+int main() 
+{
+    if (reboot_error_detected == 1) {
+        if (MBED_SUCCESS == mbed_get_reboot_error_info(&error_ctx)) {
+            printf("\nRead in reboot info\n");
+            if (error_ctx.error_status == MBED_ERROR_HARDFAULT_EXCEPTION) {
+               if (MBED_SUCCESS == mbed_get_reboot_fault_context(&fault_ctx)) {
+                   printf("\nRead in fault context info\n");
+               }
+            }
+        }
+    }
+    //main continues...
+}
+```
+
+# Tools and configuration changes
+
+### Platform configuration options for error reporting infrastruture
+
+Below is the list of new configuration options needed to configure error reporting functionality. All of these configuration options should be captured in mbed_lib.json file in platform directory.
+
+**crash-capture-enabled**
+
+Enables crash context capture when the system enters a fatal error/crash. When this is disabled it should also disable other dependent options.
+
+**fatal-error-auto-reboot-enabled**
```

Comment:
Maybe it's just me, but across the document, exception, crash, and error are used interchangably. Would prefer if only one were used, otherwise when we look back at the config options outside of the context of this PR, the config options will appear disparate.

Reply:
I see the point, the crash word comes from requirement which comprises both fatal exceptions and fatal errors. Exceptions are specifically referring to processor exceptions and errors refer to fatal errors in the system which ends up calling mbed_error() interface. Let me clarify these terminologies in assumption section. Hope that helps.